src/metaopt/worker/experiment.py

```diff
@@ -0,0 +1,396 @@
+
+        if score_handle is None:
+            selected_trial = random.sample(new_trials, 1)[0]
+        else:
+            raise NotImplementedError("scoring will be supported in the next iteration.")
+
+        selected_trial.status = 'reserved'
+        selected_trial.start_time = datetime.datetime.utcnow()
+
+        self._db.write('trials', selected_trial.to_dict(),
+                       query={'_id': selected_trial.id})
+
+        return selected_trial
+
+    def push_completed_trial(self, trial):
+        """Inform database about an evaluated `trial` with results.
+
+        :param trial: Corresponds to a successful evaluation of a particular run.
+        :type trial: `Trial`
+
+        .. note:: Change status from *reserved* to *completed*.
+        """
+        trial.end_time = datetime.datetime.utcnow()
+        trial.status = 'completed'
+        self._db.write('trials', trial.to_dict(), query={'_id': trial.id})
+
+    def register_trials(self, trials):
+        """Inform database about *new* suggested trial with specific parameter
+        values. Each of them correspond to a different possible run.
+
+        :type trials: list of `Trial`
+        """
+        stamp = datetime.datetime.utcnow()
+        for trial in trials:
+            trial.experiment = self._id
+            trial.status = 'new'
+            trial.submit_time = stamp
+        trials_dicts = list(map(lambda x: x.to_dict(), trials))
+        self._db.write('trials', trials_dicts)
+
+    def fetch_completed_trials(self):
+        """Fetch recent completed trials that this `Experiment` instance has not
+        yet seen.
+
+        .. note:: It will return only those with `Trial.end_time` after
+           `_last_fetched`, for performance reasons.
+
+        :return: list of completed `Trial` objects
+        """
+        query = dict(
+            experiment=self._id,
+            status='completed',
+            end_time={'$gte': self._last_fetched}
+            )
+        completed_trials = Trial.build(self._db.read('trials', query))
+        self._last_fetched = datetime.datetime.utcnow()
+
+        return completed_trials
+
+    @property
+    def is_done(self):
+        """Count how many trials have been completed and compare with `max_trials`.
+
+        .. note:: To be used as a terminating condition in a ``Worker``.
+        """
+        query = dict(
+            experiment=self._id,
+            status='completed'
+            )
+        num_completed_trials = self._db.count('trials', query)
+        if num_completed_trials >= self.max_trials:
+            return True
+        return False
+
+    @property
+    def configuration(self):
+        """Return a copy of an `Experiment` configuration as a dictionary."""
+        # After successful initialization, get a dict form from DB.
+        # Check :attr:`algorithms` and :attr:`refers` to see why.
+        if self._init_done:
+            return self._db.read('experiments', {'_id': self._id})[0]
```

Comment:
Oh sorry about that. Copying from `experiment.py`, lines 291-304:

```python
        self.status = 'pending'
        final_config = self.configuration  # grab dict representation of Experiment

        # Sanitize and replace some sections with objects
        self._sanitize_config()

        # If everything is alright, push new config to database
        if is_new:
            self._db.write('experiments', final_config)
            # XXX: Reminder for future DB implementations:
            # MongoDB, updates an inserted dict with _id, so should you :P
            self._id = final_config['_id']
        else:
            self._db.write('experiments', final_config, {'_id': self._id})
```

The `final_config`, pushed into the database here, is a dictionary representation of an `Experiment` object's settings. However, check that in `self._sanitize_config()`, I intend to change `algorithms` and `refers` keys to actual objects to be able to use them immediately. So `_sanitize_config` kills two birds in one stone:

1. Checks whether what is given in the config is actually valid and instantiable (by trying to instantiating them).
2. Prepares ground for actual usage of settings through functional objects.

In response, if someone likes to have a dictionary copy of an `Experiment`'s settings, it is easier to pull it from the database, instead of rebuilding it when the `configuration` property is needed.

Reply:
Sorry, I still don't understand. :blush: The dictionary ```final_config``` is not changed by ```self._sanitize_config()``` right? If yes then ```config``` from ```db.read()``` is the same as the one built in the loop at line 239. Rebuilding it might actually be less expensive than a DB call, so I still don't understand why the ```if init: return db.read()``` is useful.

Reply:
`final_config` is a local variable which contains the dict form of the experiment. This will be pushed to the database.

Attributes of an `Experiment` on the other hand will not remain in the dict form when `_sanitize_config` is called. So, for instance, `algorithms` is going to be replaced by the algorithm object. This means that if we were to iterate over the attributes in order to create the dict form, `algorithms` is going to have the object itself instead of the object's dict form. So to avoid these conversions, I fetch from the db, which has already saved the final config (and it is in dict form).

Reply:
It's getting clearer now thanks. What if the config is a yaml file with instantiation keys (```obj!:module.foo```), the config passed to ```Experiment``` will contain objects other than dicts and strings. Will we do 2 pass initialisations?
1. [outside experiment] Yaml -> Dicts (instantiation keys are kept)
2. [inside experiment._sanitize_config] Dicts -> Objects  (instantiation keys are converted to objects)

Also, what would you think of setting ```self._configuration``` at configuration time? It would be less expensive to fetch than ```db.read()``` and I can't believe that configuration dictionaries will take a large amount of memory.

Reply:
Nice point about the yaml instantiation keys!!

Ok let's raise an issue for this too! `_sanitize_config` should bring `Experiment` to the usable form. After your point, I think that the easiest solution would be recreating the dictionary form from possible algorithm objects in `configuration` property. I need to write some code for an `AbstractAlgorithm` first I think in order to visualize exactly how this is going to be done. For now let's assume that we do not have instantiation keys from YAML, all configuration files are dictionaries.

So I think about:
1. [outside experiment] just use YAML -> results to a dictionary of things
2. [inside experiment._sanitize_config] check if objects already instantiated are in fact `AbstractAlgorithms`. If not raise; if they are ok. Also, convert rest dicts to objects when applicable through the `Factory`.
3. [inside experiment.configure] Convert possible objects to a dict descriptor. For algorithms this will be a class method, I was thinking to do this anyway to (auto)generate example configuration files from `AbstractAlgorithm` implementations (later later :P). But for referred experiments I do not know how this is going to be.

I propose to let it as it is for now. I will proceed by implementing interfaces for algorithms and wrapping one example. Let's revisit it later when we have an image of what the algorithm will do.

Reply:
> For now let's assume that we do not have instantiation keys from YAML, all configuration files are dictionaries.

I agree

Just to make sure I understand well, do you propose that algorithms also have a ```configuration``` property? I think that's a good idea and would make it easy to rebuilt configuration dictionary within the ```Experiment``` object. Any other configuration object within ```Experiment``` could also have this ```configuration``` property.

For now, what would you think of using ```self._configuration``` rather than ```db.read()```?

Reply:
I can imagine a default configuration property in the `AbstractAlgorithm` that returns a dictionary form of its specification. Something like:

```python
dict_form = dict()
for attrname in self.__slots__:
    dict_form[attrname] = getattr(instance, attrname, self.attrs[attrname]['default_value'])
    if isinstance(dict_form[attrname], AbstractAlgorithm):
        dict_form[attrname] = dict_form[attrname].configuration

return {self.CLASS_ID: dict_form}
```

I will leave `self._db.read(...)` for now. I want to move on to optimizers' interface etc

Reply:
Alright, we'll have to revisit this part anyway.